### ISSUE TITLE
Question: Which aws permissions does the IAM role need exactly?

### ISSUE THREAD

**Issue by DerFels** [nerdswords/yet-another-cloudwatch-exporter#440]
Hi,
I'm testing YACE version 0.30.0-alpha in a multi AWS Account setup. However, I cannot get it to work :(

I'm running YACE in a docker container and configured an AWS access key + region in an .env file which is specified in the docker-compose file.
### .env file
```
AWS_ACCESS_KEY_ID=ABCDABCD...
AWS_SECRET_ACCESS_KEY=supersecretkey
AWS_DEFAULT_REGION=eu-central-1
```

### docker-compose file
```
version: '2'

services:
  cloudwatch_exporter:
    image: ghcr.io/nerdswords/yet-another-cloudwatch-exporter:v0.30.0-alpha
    env_file:
      - .env
    build: ./
    ports:
      - 5000:5000

```
and I copy a config file into the container into the /tmp folder
### config.yml
```
apiVersion: v1alpha1
static:
  - namespace: AWS/ES
    name: Elasticsearch
    regions:
      - eu-central-1
    roles:
      - roleArn: "arn:aws:iam:111111111111:role/Prometheus" # Account 1
      - roleArn: "arn:aws:iam:222222222222:role/Prometheus" # Account 2
    dimensions:
     - name: ClientId
     - name: DomainName
    metrics:
      - name: ClusterStatus.green
        statistics:
        - Maximum
        period: 60
        length: 300
```

The AWS Keys I specify in the .env file, belong to an IAM User within AWS Account 1, which has the permission to assume the IAM Role called "Prometheus" in every one of our AWS Accounts
```
...
{
            "Sid": "",
            "Effect": "Allow",
            "Action": "sts:AssumeRole",
            "Resource": [
                "arn:aws:iam::111111:role/Prometheus",
                "arn:aws:iam::22222:role/Prometheus",
                "arn:aws:iam::333333:role/Prometheus",
                "arn:aws:iam::44444:role/Prometheus",
                "arn:aws:iam::55555:role/Prometheus",
                "arn:aws:iam::666666:role/Prometheus",
                "arn:aws:iam::77777:role/Prometheus"
            ]
        }
```

The Prometheus IAM Role has the permissions to get the Cloudwatch logs.
In theory, all should work fine, but in the YACE Logs I can only see

```
cloudwatch_exporter_1  | {"level":"warning","msg":"Unable to get metric statistics due to AccessDenied: User: arn:aws:iam::111111111:user/prometheus is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam:222222222:role/Prometheus\n\tstatus code: 403, request id: 3f79ead6-bbec-411e-97bc-1e915f30e4d5","time":"2021-09-13T12:32:13Z"}
cloudwatch_exporter_1  | {"level":"warning","msg":"Unable to get metric statistics due to AccessDenied: User: arn:aws:iam::111111111:user/prometheus is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam:111111111:role/Prometheus\n\tstatus code: 403, request id: 8d2bb592-c314-4a42-84b6-057a46c40472","time":"2021-09-13T12:32:14Z"}
```

So, according to YACE, the IAM user is neither authorized to assume the role within the same account, nor in the other AWS Account.
However, when trying to assume the role as the user myself on the AWSCLI, it works fine.

Any help is greatly appreciated :)

**Comment by thomaspeitz** [comment]
Will verify this setup on Thuersday. Thanks for the nice description.

If you could by any chance verify in your cloudtrail what happens, this could help me a lot.
Maybe there we find a hint what is actually happening.

Thank you!

**Comment by DerFels** [comment]
Thanks for looking into this.
I did another test today and checked Cloudtrail. I saw the following:
- Requests to other accounts were successful (Role successfully assumed)
- Request to assume the Role in the same account that the IAM User lives in -> Access Denied

I played around and even set the Trust Relationship of the IAM Role to "*", but it didn't change a thing.
Sidenote: I am using the prometheus-cloudwatch-exporter in parallel and it can retrieve the metrics from every account without problems.

Another thing I noticed in the Cloudtrails is that the YACE requests always specify: region=us-east-1. Even though I configured the region eu-central-1 via Env vars (default_region + region) and via the config.yml. But this doesn't seem to have any impact.

One more thing I observed:
When using a `discovery` config for YACE, it doesn't seem to use the `-scraping-interval` parameter. It worked fine with the `static` config.

**Comment by thomaspeitz** [comment]
Thanks a lot that helps me. Will verify this behaviour and try to fix it.

The discovery is a new issue which you can add as well (with more informations) if you would like to. Keeping it out of this issue.

**Comment by DerFels** [comment]
Thanks, I will open another issue for the static vs. discovery topic.
Just wanted to add this output which I got today from the `static` configuration
> cloudwatch_exporter_1  | {"level":"info","msg":"Parse config..","time":"2021-09-21T12:00:11Z"}
cloudwatch_exporter_1  | {"level":"info","msg":"Startup completed","time":"2021-09-21T12:00:11Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"Starting scraping async","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"info","msg":"Couldn't get account Id for role arn:aws:iam:111111:role/Prometheus: AccessDenied: User: arn:aws:iam::1111111:user/prometheus is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam:11111111:role/Prometheus\n\tstatus code: 403, request id: f808b6ca-28f1-43a4-8868-321e67df2350\n","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"CLI helper - aws cloudwatch get-metric-statistics --metric-name ClusterStatus.green --dimensions Name=ClientId,Value= Name=DomainName,Value=  --namespace AWS/ES --statistics Maximum --period 60 --start-time 2021-09-21T11:55:21Z --end-time 2021-09-21T12:00:21Z","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"{\n  Dimensions: [{\n      Name: \"ClientId\",\n      Value: \"\"\n    },{\n      Name: \"DomainName\",\n      Value: \"\"\n    }],\n  EndTime: 2021-09-21 12:00:21.936246783 +0000 UTC m=+10.945927379,\n  MetricName: \"ClusterStatus.green\",\n  Namespace: \"AWS/ES\",\n  Period: 60,\n  StartTime: 2021-09-21 11:55:21.936247144 +0000 UTC m=-289.054072261,\n  Statistics: [\"Maximum\"]\n}","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"{\n  Dimensions: [{\n      Name: \"ClientId\",\n      Value: \"\"\n    },{\n      Name: \"DomainName\",\n      Value: \"\"\n    }],\n  EndTime: 2021-09-21 12:00:21.936246783 +0000 UTC m=+10.945927379,\n  MetricName: \"ClusterStatus.green\",\n  Namespace: \"AWS/ES\",\n  Period: 60,\n  StartTime: 2021-09-21 11:55:21.936247144 +0000 UTC m=-289.054072261,\n  Statistics: [\"Maximum\"]\n}","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"{\n\n}","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"warning","msg":"Unable to get metric statistics due to InvalidParameter: 2 validation error(s) found.\n- minimum field size of 1, GetMetricStatisticsInput.Dimensions[0].Value.\n- minimum field size of 1, GetMetricStatisticsInput.Dimensions[1].Value.\n","time":"2021-09-21T12:00:21Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"Metrics scraped.","time":"2021-09-21T12:00:21Z"}
Killing cloudwatch-exporter_cloudwatch_exporter_1    ... done
ERROR: 2

So I thought I might have to specify a Value for at leat one dimension. But when I tried it like this
```
dimensions:
      - name: ClientId
      - name: DomainName
        value: monitoring-es-cluster
```
it failed with the same error
`Unable to get metric statistics due to InvalidParameter: 2 validation error(s) found.\n- minimum field size of 1, GetMetricStatisticsInput.Dimensions[0].Value.\n- minimum field size of 1, GetMetricStatisticsInput.Dimensions[1].Value`

And just for the sake of completeness: with a `discovery` config, the aws-cli command is missing in the debug output.

> cloudwatch_exporter_1  | {"level":"info","msg":"Parse config..","time":"2021-09-21T12:02:49Z"}
cloudwatch_exporter_1  | {"level":"info","msg":"Startup completed","time":"2021-09-21T12:02:49Z"}
cloudwatch_exporter_1  | {"level":"debug","msg":"Starting scraping async","time":"2021-09-21T12:07:48Z"}
cloudwatch_exporter_1  | {"level":"info","msg":"Couldn't get account Id for role arn:aws:iam:1111111:role/Prometheus: AccessDenied: User: arn:aws:iam::11111111:user/prometheus is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam:11111111:role/Prometheus\n\tstatus code: 403, request id: 861b4816-e5e3-4a1c-95f7-7cefce318940\n","time":"2021-09-21T12:07:49Z"}
cloudwatch_exporter_1  | panic: runtime error: invalid memory address or nil pointer dereference
cloudwatch_exporter_1  | [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd01aa4]
cloudwatch_exporter_1  | 
cloudwatch_exporter_1  | goroutine 8 [running]:
cloudwatch_exporter_1  | github.com/ivx/yet-another-cloudwatch-exporter/pkg.scrapeAwsData.func1(0x0, {0xc0001badc0, 0xc}, {{0xc000296210, 0xc00003af90}, {0x0, 0xc00003afa0}})
cloudwatch_exporter_1  |        /opt/pkg/abstract.go:40 +0x424
cloudwatch_exporter_1  | created by github.com/ivx/yet-another-cloudwatch-exporter/pkg.scrapeAwsData
cloudwatch_exporter_1  |        /opt/pkg/abstract.go:25 +0x886
cloudwatch-exporter_cloudwatch_exporter_1 exited with code 2

**Comment by thomaspeitz** [comment]
@DerFels - Could you please try the following (See instead of iam `:`, you use `::`)
```
    roles:
      - roleArn: "arn:aws:iam::111111111111:role/Prometheus" # Account 1
      - roleArn: "arn:aws:iam::222222222222:role/Prometheus" # Account 2
```

I think the docs are not correct here.

Build 40 minutes a test setup and debugged code till I realized it :D
For me it works like that

**Comment by DerFels** [comment]
Oh man :D
Yes I can confirm that when using `iam::` it works perfectly.

Thanks Thomas!
